### ISSUE TITLE
chef_vault_item by default provides encrypted databag in production when decryption or key not found

### ISSUE THREAD

**Issue by jaybocc2** [opscode-cookbooks/chef-vault#26]
this seems to be a regression and an exception is much more useful/easy to handle than a raw encrypted databag.

This issue was introduced in d4a3bdfa545c761ce341ad972999ae72c17ab846 when node['dev_mode'] condition was removed.

perhaps others are ok with this but in my opinion its much better practice to explicitly enable a development workflow rather than default to poor behavior in production.

**Comment by mivok** [comment]
d4a3bdf was deliberately done to allow cookbooks to use chef_vault_item to not fail when an unencrypted data bag was used. This isn't a 'development workflow' but more allowing cookbooks to be compatible with people who don't use chef vault at all as well as those who do. It's unfortunate that the code looks very similar to the original 'dev_mode' flag, but development is only one use case for this change.

Do you still see the same behavior if catching SecretDecryption is dropped, but catching KeysNotFound is left in? The KeysNotFound exception is only raised when the vault_item_keys data bag item isn't there, implying that what you're trying to access isn't a vault item after all, but a normal data bag. It's not raised when chef vault can't find a key to decrypt with.

I believe this issue was introduced in #14 and reverting that may cause the desired behavior in all cases. At the moment I'm looking into what situation #14 fixed to see if this needs to be worked around in a different way. /cc @someara

**Comment by kbaltrinic** [comment]
I am not certain that I am 100% following the above conversation so if this is a separate issue, I'll open a new ticket.  That said, I am seeing what I think is the same problem wherein I have items in a vault  and a node attempts to access them before the vault has been refreshed to permit that server (typically happening on a `knife bootstrap` activity.  As a result of not being in the authorized list for the item yet, the attempt to get the vault item fails and falls back to the development fallback of trying to grab a data bag of the same name.  This succeeds in so far as a vault is a data bag.  What happens next is the that chef run fails with an error that an object was received when expecting a string, because what would have been a plain text string in an normal data bag is an object in a vault data bag.  

To my mind what is needed here is some detection logic in the development fallback code [here](https://github.com/opscode-cookbooks/chef-vault/blob/1fb3fb7b95ce04ab54560d883aee9c1113bbe8ba/libraries/chef_vault_item.rb#L43) that if the value retrieve from the data bag is an object having the same keys as a vault encrypted item would have, then an the `ChefVault::Exceptions::KeysNotFound` exception should be re-raised.

Thoughts?

**Comment by mivok** [comment]
Hi Ken. This is the same issue that you're experiencing.

We currently have #28 open which fixes this without needing to catch KeysNotFound nor SecretDecryption exceptions. Once that is merged, it should fix your problem.

**Comment by kbaltrinic** [comment]
So I looked at that after I posted my comments (backwards I know). i\It pretty much looks like what I was recommending.  Approach to auto-detecting vaults vs data_bags is a bit different but makes better sense than mine.

**Comment by jtimberman** [comment]
#28 is merged. Is this still an issue?

**Comment by jaybocc2** [comment]
This looks like this does exactly what is needed to resolve this issue.  I am unable to roll to production before weeks end but so far so good.
